include/utilities/cartesian.hpp

```diff
@@ -0,0 +1,39 @@
+#ifndef NGEN_UTILITIES_CARTESIAN_HPP
+#define NGEN_UTILITIES_CARTESIAN_HPP
+
+#include <vector>
+
+#include "span.hpp"
+
+namespace ngen {
+
+/**
+ * @brief 
```

Comment:
Without scribbling out a lot of algebra, I can't directly work out what this function does. Please document it more extensively.

I'm also concerned about the mixed iterative/recursive structure, as well as the combinatorial size of output that it implies.

Style-wise, it's writing to elements of `index` and extending `output`. It looks like `index` is just a bit of scratch memory for use during the overall operation. Given that, it would be preferable to rename this routine something like `cartesian_indices_helper`, and have a `cartesian_indices` routine that allocates `index` internally and passes it down, so that the caller doesn't need to. That split could potentially also be an opportunity address some of my concerns about the mixed iterative/recursive structure.

Reply:
Finally, as a big alternative to actually materializing a container that holds all of the indices, you could make more of a generator object the can spit out the indices one by one, without first generating and storing all of them.

Reply:
> This comment is just for context

This function takes in a `shape` span, (i.e. `{3, 1, 2}`), a buffer `index`, the current `dimension`, and the actual `output`.

In general what this does is generates all the possible indices that can occur from the given `shape`. For example, calling `cartesian_indices` with the above shape would result in:
```
output = {
  { 0, 0, 0 },
  { 0, 0, 1 },
  { 1, 0, 0 },
  { 1, 0, 1 },
  { 2, 0, 0 },
  { 2, 0, 1 }
}
```

I agree that there's probably a better way, and a generator object may be the way to go (especially as the `mdframe` contents get larger)

Reply:
After working on this for a bit, I don't think there's a good solution for splitting the iteration and recursive components (at least not from what I could see). Due to the cartesian product, there will be a need for iteration (in this case, we iterate to make recursive calls), and the recursion makes it succinct. Since this is an output handler (and _should_ only be called at the very end of a run), I'm not sure that refactoring the inner details is worth it.

> One thing to note about `output` as well is that its size is known before this function is ever called (since it's just the product of the elements of `shape`). So, in theory, we could handle it outside of this if it is a concern.

However, modifying it so that only the shape and output vector of vectors are needed as inputs is possible (which I'll add onto the next push).